From 0a11629c915f61df798919db51a18ffe4649cb65 Mon Sep 17 00:00:00 2001
From: Daniel Henrique Barboza <dbarboza@ventanamicro.com>
Date: Thu, 14 Mar 2024 14:57:04 -0300
Subject: [PATCH 12/12] target/riscv/vector_helper.c: optimize loops in ldst
 helpers

Change the for loops in ldst helpers to do a single increment in the
counter, and assign it env->vstart, to avoid re-reading from vstart
every time.

Suggested-by: Richard Henderson <richard.henderson@linaro.org>
Signed-off-by: Daniel Henrique Barboza <dbarboza@ventanamicro.com>
Reviewed-by: Alistair Francis <alistair.francis@wdc.com>
Reviewed-by: Richard Henderson <richard.henderson@linaro.org>
Message-ID: <20240314175704.478276-11-dbarboza@ventanamicro.com>
Signed-off-by: Alistair Francis <alistair.francis@wdc.com>
Origin: backported, https://gitlab.com/qemu-project/qemu/-/commit/0a11629c
Bug: https://gitlab.com/qemu-project/qemu/-/issues/1976
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/qemu/+bug/2095169
---
 target/riscv/vector_helper.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

--- a/target/riscv/vector_helper.c
+++ b/target/riscv/vector_helper.c
@@ -198,7 +198,7 @@

     VSTART_CHECK_EARLY_EXIT(env);

-    for (i = env->vstart; i < env->vl; i++, env->vstart++) {
+    for (i = env->vstart; i < env->vl; env->vstart = ++i) {
         k = 0;
         while (k < nf) {
             if (!vm && !vext_elem_mask(v0, i)) {
@@ -266,7 +266,7 @@
     VSTART_CHECK_EARLY_EXIT(env);

     /* load bytes from guest memory */
-    for (i = env->vstart; i < evl; i++, env->vstart++) {
+    for (i = env->vstart; i < evl; env->vstart = ++i) {
         k = 0;
         while (k < nf) {
             target_ulong addr = base + ((i * nf + k) << log2_esz);
@@ -382,7 +382,7 @@
     VSTART_CHECK_EARLY_EXIT(env);

     /* load bytes from guest memory */
-    for (i = env->vstart; i < env->vl; i++, env->vstart++) {
+    for (i = env->vstart; i < env->vl; env->vstart = ++i) {
         k = 0;
         while (k < nf) {
             if (!vm && !vext_elem_mask(v0, i)) {
